[17.0][MIG] fastapi: Migration to 17.0

nguyenminhchien opened the pull request:
Ref: BSRD-708

Depends on:
- [x] https://github.com/OCA/web-api/pull/31

lmignon committed:
[FIX] fastapi: In test class don't override overrides

If you pass an override of the authenticated_partner_impl to the _create_test_client method, don't override it by a default one

OCA-git-bot committed:
[BOT] post-merge updates

sbidoul committed:
[FIX] fastapi: close cursor after rollback

This is to ensure that `retrying` in `service/model.py` does not try to flush.

OCA-git-bot committed:
[BOT] post-merge updates

sbidoul committed:
Revert "[FIX] fastapi: close cursor after rollback"

This reverts commit 8be7fbb0fda92b46e4913176185fb4e19ab319be.

OCA-git-bot committed:
[BOT] post-merge updates

nguyenminhchien commented:
The check failed because of the issue [#141747](https://github.com/odoo/odoo/issues/141747)

mymage committed:
Added translation using Weblate (Italian)

hunghn commented:
Thank you, Mr. Chien, for your contribution. It works for me.

AnizR committed:
[16.0][FIX] fastapi: always add authenticate_partner_id in context when evaluating ir rules

This fix is needed since a modificiation of ir.rule that checks the domain when creating/modifying ir rules
The solution is to set authenticate_partner_id to False when it is not present in context

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (41 of 41 strings)

Translation: rest-framework-16.0/rest-framework-16.0-fastapi
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi/it/

OCA-git-bot committed:
[BOT] post-merge updates

lmignon reviewed (commented):
@nguyenminhchien Thank you for the migration. Con you remove the dependency on `odoo-addon-endpoint-route-handler @ git+https://github.com/OCA/web-api.git@refs/pull/31/head#subdirectory=endpoint_route_handler` This one is merged.

nguyenminhchien commented:
> @nguyenminhchien Thank you for the migration. Con you remove the dependency on `odoo-addon-endpoint-route-handler @ git+https://github.com/OCA/web-api.git@refs/pull/31/head#subdirectory=endpoint_route_handler` This one is merged.

Removed, thanks.

gurneyalex commented:
/ocabot migration fastapi

gurneyalex commented:
@lmignon @sbidoul it seems that Odoo 17 has issues with the addons having a 16.0.x.y.z version, even if they are installable=False

cf. https://github.com/OCA/rest-framework/actions/runs/7634588421/job/20798693591?pr=409#step:7:74

Is this something you prefer reporting as an issue to Odoo or handle in the test scripts or process with a commit updating the versions of the uninstallable addons in the repo?

lmignon commented:
> @lmignon @sbidoul it seems that Odoo 17 has issues with the addons having a 16.0.x.y.z version, even if they are installable=False
> 
> cf. https://github.com/OCA/rest-framework/actions/runs/7634588421/job/20798693591?pr=409#step:7:74
> 
> Is this something you prefer reporting as an issue to Odoo or handle in the test scripts or process with a commit updating the versions of the uninstallable addons in the repo?

An issue is open on odoo regarding this problem https://github.com/odoo/odoo/issues/141747

rven commented:
After an update of Odoo the following error is raised on the fastapi module
> odoo.tools.convert.ParseError: while parsing /opt/odoo/parts/oca_pull/fastapi/security/ir_rule+acl.xml:37
> 2002Invalid domain: <class 'NameError'>: "name 'authenticated_partner_id' is not defined" while evaluating
> 2003" ['|', ('user_id', '=', user.id), ('id', '=', authenticated_partner_id)]"
To fix this, the following code needs to be changed in `fastapi/models/ir_rule.py`
```python
    @api.model
    def _eval_context(self):
        ctx = super()._eval_context()
        if "authenticated_partner_id" in self.env.context:
            ctx["authenticated_partner_id"] = self.env.context[
                "authenticated_partner_id"
            ]
        return ctx
```
Should be changed to this

```python
    @api.model
    def _eval_context(self):
        ctx = super()._eval_context()
        ctx["authenticated_partner_id"] = self.env.context.get("authenticated_partner_id")
        return ctx
```

sbidoul commented:
@rven this looks like https://github.com/OCA/rest-framework/pull/410

nguyenminhchien commented:
> @rven this looks like #410

@rven @sbidoul I have picked https://github.com/OCA/rest-framework/pull/410

lmignon committed:
[FIX] fastapi: Compatibility with latest Odoo

From https://github.com/odoo/odoo/commit/cb1d057dcab28cb0b0487244ba99231ee292502e, the orignal werkzeug request is wrapped in a dedicated class to keep under control the attributes developers use. This change add code to make the fastapi addon working with version including this last change and prior version

refs #414

ahmed-aly-aut commented:
@nguyenminhchien this pull request is now needed too #416

nguyenminhchien commented:
> @nguyenminhchien this pull request is now needed too #416

I picked this one.

lmignon commented:
/ocabot merge nobump

lmignon reviewed (approved):
Code review only

dreispt commented:
/ocabot merge nobump

dreispt commented:
/ocabot merge nobump

astirpe commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -2,9 +2,10 @@
 from functools import partial
 from itertools import chain
-from typing import Any, Awaitable, Callable, Dict, List, Tuple, Type, Union
+from typing import Any
```
Where these `Dict, List, Tuple, Type, Union` removed accidentally by ruff, or was it intentional?

nguyenminhchien replied:
Hi, they were deprecated since version 3.9 ( i read the information here: https://docs.python.org/3/library/typing.html)
![image](https://github.com/OCA/rest-framework/assets/8724793/0bd7b555-7df8-4ddb-adf8-bcef5bce3ee2)

astirpe replied:
You are right, thank you!

dreispt commented:
/ocabot merge nobump

ahmed-aly-aut commented:
can someone approve?

lmignon commented on fastapi/models/fastapi_endpoint.py:
```diff
@@ -2,9 +2,10 @@
 # License LGPL-3.0 or later (http://www.gnu.org/licenses/LGPL).
 
 import logging
+from collections.abc import Awaitable, Callable
```
Awaitable ans Callable should be imported from typing

nguyenminhchien replied:
Hi, they were deprecated since version 3.9 ( i read the information here: https://docs.python.org/3/library/typing.html).
![image](https://github.com/OCA/rest-framework/assets/8724793/4f489757-56c1-43e5-af8b-23b1773ac79d)

nguyenminhchien committed:
[MIG] fastapi: Migration to 17.0

nilshamerlinck commented:
Hi @dreispt could you please try again your merge bot command? I believe it wasn't going through previously because of an old pending runboat build, but this has now been fixed thanks to Chien's rebase

lmignon commented:
/ocabot merge nobump

nilshamerlinck commented:
>  Hi @dreispt could you please try again your merge bot command? I believe it wasn't going through previously because of an old pending runboat build, but this has now been fixed thanks to Chien's rebase

Yikes, wrong hypothesis :sweat_smile: 

Hi @sbidoul the 17.0 branch of this repo is based on 16.0 and [some manual changes](https://github.com/OCA/rest-framework/commit/c8172049ce6cd174c2cf2e9d0707f8a67bae4ce8)

Do you think it could now be initialized "normally" through copier?

sbidoul commented on fastapi/__manifest__.py:
```diff
@@ -30,5 +30,5 @@
     },
     "development_status": "Beta",
-    "installable": False,
+    # "installable": False,
```
```suggestion
    "installable": True,
```

Or remove the line.

sbidoul commented on fastapi/schemas.py:
```diff
@@ -37,8 +37,8 @@ def total(self, value: int):
 class Paging(BaseModel):
-    limit: Optional[int] = None
-    offset: Optional[int] = None
+    limit: Optional[int] = None  # noqa: UP007
+    offset: Optional[int] = None  # noqa: UP007
```
```suggestion
    limit: int | None
    offset: int | None
```

Would this work?

nguyenminhchien replied:
yes indeed

sbidoul reviewed (commented):
A couple of comments "en passant".

There are a few missing commits from the 16.0 branch. Would be nice to pick them up before merging.

I'll look at why the bot does not merge.

sbidoul commented:
Hm, weird. Trying again.

/ocabot merge nobump

sbidoul commented:
Ah I understand https://github.com/OCA/rest-framework/pull/409#issuecomment-2132153109 now. The base branch CI was sill configured for 16.0.  I ran `copier update --trust`, answered the questions for 17.0 and rebased.

lmignon commented:
/ocabot merge nobump

lmignon commented:
/ocabot merge nobump
